File: leetcode/0875-KokoEatingBananas/solution.cpp
class Solution {
    private:
        bool isOk(int val, int h, vector<int>& piles) {
            auto sum = 0;
            for (const auto& p : piles) {
                sum += p/val + (p % val > 0 ? 1: 0); 
                if (sum > h) {
                    return false;
                }
            }
            return (sum <= h);
        }
        int calculateMinSpeed(int start, int end, int h, vector<int>& piles) {
            if (start >= end) {
                return end;
            }
            const auto mid = (start + end)/2;
    
            if (isOk(mid, h, piles)) {
                return calculateMinSpeed(start, mid, h, piles);
            } else {
                return calculateMinSpeed(mid+1, end, h, piles);
            }
        }    
    public:
        int minEatingSpeed(vector<int>& piles, int h) {
            const auto maxPile = *max_element(piles.begin(), piles.end());
            return calculateMinSpeed(1, maxPile, h, piles);
            
        }
    };
    

File: leetcode/0875-KokoEatingBananas/solution.go
func max(piles []int) int {
	max := piles[0]
	for _, val := range piles {
		if val > max {
			max = val
		}
	}
	return max
}

func canKokoEatAll(eatRate int, piles []int, h int) bool {
	sum := 0
	for _, p := range piles {
		sum += p / eatRate
		if p%eatRate > 0 {
			sum += 1
		}
		if sum > h {
			return false
		}
	}

	return sum <= h
}

func binarySearch(start int, end int, piles []int, h int) int {
	if start >= end {
		return end
	}

	mid := (start + end) / 2

	if canKokoEatAll(mid, piles, h) {
		return binarySearch(start, mid, piles, h)
	} else {
		return binarySearch(mid+1, end, piles, h)
	}
}

func minEatingSpeed(piles []int, h int) int {
	maxPile := max(piles)
	return binarySearch(1, maxPile, piles, h)

}
